trace/trace.go

```diff
@@ -123,73 +107,83 @@ func WithSpan(parent context.Context, s *Span) context.Context {
+func StartSpanWithOptions(ctx context.Context, name string, o StartOptions) (context.Context, *Span) {
 	parentSpan, _ := ctx.Value(contextKey{}).(*Span)
-	return WithSpan(ctx, parentSpan.StartSpanWithOptions(name, o))
+	span := NewSpan(name, parentSpan, o)
+	return WithSpan(ctx, span), span
 }
 
-// StartSpanWithRemoteParent starts a new child span with the given parent SpanContext.
-//
-// If there is an existing span in ctx, it is ignored -- the returned Span is a
-// child of the span specified by parent.
-func StartSpanWithRemoteParent(ctx context.Context, name string, parent SpanContext, o StartSpanOptions) context.Context {
-	return WithSpan(ctx, NewSpanWithRemoteParent(name, parent, o))
-}
-
-// StartSpan starts a new child span.
+// EndSpan ends the current span.
 //
-// If s is nil, creates a new trace and span, like the function NewSpan.
-func (s *Span) StartSpan(name string) *Span {
-	return s.StartSpanWithOptions(name, StartSpanOptions{})
+// The context passed to EndSpan will still refer to the now-ended span, so any
+// code that adds more information to it, like SetSpanStatus, should be called
+// before EndSpan.
+func EndSpan(ctx context.Context) {
+	s, ok := ctx.Value(contextKey{}).(*Span)
+	if !ok {
+		return
+	}
+	s.End(EndOptions{Tags: tag.FromContext(ctx)})
 }
 
-// StartSpanWithOptions starts a new child span using the given options.
+// StartSpanWithRemoteParent starts a new child span with the given parent SpanContext.
 //
-// If s is nil, creates a new trace and span, like the function NewSpan.
-func (s *Span) StartSpanWithOptions(name string, o StartSpanOptions) *Span {
-	if s != nil {
-		return startSpanInternal(name, true, s.spanContext, false, o)
-	}
-	return startSpanInternal(name, false, SpanContext{}, false, o)
+// If there is an existing span in ctx, it is ignored -- the returned Span is a
+// child of the span specified by parent.
+func StartSpanWithRemoteParent(ctx context.Context, name string, parent SpanContext, o StartOptions) (context.Context, *Span) {
+	span := NewSpanWithRemoteParent(name, parent, o)
+	return WithSpan(ctx, span), span
 }
 
 // NewSpan returns a new span.
 //
-// The returned span has no parent span; a new trace ID will be created for it.
-func NewSpan(name string, o StartSpanOptions) *Span {
-	return startSpanInternal(name, false, SpanContext{}, false, o)
+// If parent is not nil, created span will be a child of the parent.
+func NewSpan(name string, parent *Span, o StartOptions) *Span {
+	hasParent := false
+	var parentSpanContext SpanContext
+	if parent != nil {
+		hasParent = true
+		parentSpanContext = parent.SpanContext()
+	}
+	return startSpanInternal(name, hasParent, parentSpanContext, false, o)
 }
 
 // NewSpanWithRemoteParent returns a new span with the given parent SpanContext.
-func NewSpanWithRemoteParent(name string, parent SpanContext, o StartSpanOptions) *Span {
+func NewSpanWithRemoteParent(name string, parent SpanContext, o StartOptions) *Span {
```

Comment:
Can we just call this NewSpan and remove the above NewSpan? Caller can always explicitly retrieve the SpanContext from the parent span and then call this method.

Reply:
We cannot remove this because NewSpanWithRemoteParent inherits the start options from the parent child.